lib/solidus_support/engine_extensions.rb

```diff
@@ -46,14 +46,38 @@ def load_solidus_subscribers_from(path)
         end
       end
 
-      # Loads decorator files.
+      # Loads decorators.
       #
       # This is needed since they are never explicitly referenced in the application code and
-      # won't be loaded by default. We need them to be executed regardless in order to decorate
-      # existing classes.
-      def load_solidus_decorators_from(path)
-        path.glob('**/*.rb') do |decorator_path|
-          require_dependency(decorator_path)
+      # won't be loaded by default. We need them to be executed whenever the decorated class is reloaded.
+      def load_solidus_decorators_from(base_path)
+        # Matches e.g. "Spree::Order.prepend"
+        decorated_class_pattern = /(?<decorated_class>[A-Z][a-zA-Z:]+)(\.prepend[\s(])/
+        # This will be Zeitwerk.
+        autoloader = Rails.autoloaders.main
+
+        base_path.glob('**/*.rb') do |path|
+          # Match all the classes that are prepended in the file
+          matches = File.open(path).read.scan(decorated_class_pattern).flatten
+
+          # Don't do a thing if there's no prepending.
+          next unless matches.present?
+
+          # For each unique match, make sure we load the decorator when the base class is loaded
+          matches.uniq.each do |decorated_class|
+            # Zeitwerk tells us which constant it expects a file to provide.
+            decorator_constant = autoloader.cpath_expected_at(path)
+
+            # Sprinkle some debugging.
+            Rails.logger.debug("Preparing to autoload #{decorated_class} with #{decorator_constant}")
+
+            # For every class name being autoloaded, we can add a hook to load the decorator when the base class is loaded.
+            # Multiple hooks are no problem, as long as all decorators are namespaced appropriately.
+            autoloader.on_load(decorated_class) do |base|
+              Rails.logger.debug("Loading #{decorator_constant} in order to modify #{base}")
+              decorator_constant.constantize
```

Comment:
Is there a Zeitwerk way to constantize or "this is the way"?

Reply:
It seems there is `on_setup` https://github.com/fxn/zeitwerk/blob/main/CHANGELOG.md?plain=1#L322-L332

Can we try that?

Reply:
Reading the Zeitwerk README  

https://github.com/fxn/zeitwerk#eager-load-directories

Can we simply append `root.join("app/decorators")` to `loader.eager_load_dir` and get rid of all this custom code. I am pretty sure Zeitwerk is able to do all the heavy lifting and deal with edge case for us, no?

Reply:
We could do that, and that would reimplement the `load` stuff we've been doing up to now. However: Decorators are "contagious" in  that when they are loaded, they also trigger autoloading of the classes they reference, along with all the classes mentioned in those classes. What this PR does is help keep us completely away from autoloadable code during startup.